app/src/github_runner_image_builder/openstack_builder.py

```diff
@@ -0,0 +1,728 @@
+    except StopIteration as exc:
+        logger.exception("Unable to determine cloud to use from clouds.yaml files.")
+        raise github_runner_image_builder.errors.CloudsYAMLError(
+            "Unable to determine cloud to use from clouds.yaml files. "
+            "Please check that clouds.yaml exists."
+        ) from exc
+    try:
+        clouds_yaml = yaml.safe_load(clouds_yaml_path.read_text(encoding="utf-8"))
+        cloud: str = list(clouds_yaml["clouds"].keys())[0]
+    except (TypeError, yaml.error.YAMLError, KeyError, IndexError) as exc:
+        logger.exception("Invalid clouds.yaml contents.")
+        raise github_runner_image_builder.errors.CloudsYAMLError("Invalid clouds.yaml.") from exc
+    return cloud
+
+
+def initialize(arch: Arch, cloud_name: str, prefix: str) -> None:
+    """Initialize the OpenStack external image builder.
+
+    Upload ubuntu base images to openstack to use as builder base. This is a separate method to
+    mitigate race conditions from happening during parallel runs (multiprocess) of the image
+    builder, by creating shared resources beforehand.
+
+    Args:
+        arch: The architecture of the image to seed.
+        cloud_name: The cloud to use from the clouds.yaml file.
+        prefix: The prefix to use for OpenStack resource names.
+    """
+    logger.info("Initializing external builder.")
+    logger.info("Downloading Jammy image.")
+    jammy_image_path = cloud_image.download_and_validate_image(
+        arch=arch, base_image=BaseImage.JAMMY
+    )
+    logger.info("Downloading Noble image.")
+    noble_image_path = cloud_image.download_and_validate_image(
+        arch=arch, base_image=BaseImage.NOBLE
+    )
+    logger.info("Uploading Jammy image.")
+    store.upload_image(
+        arch=arch,
+        cloud_name=cloud_name,
+        image_name=_get_base_image_name(arch=arch, base=BaseImage.JAMMY, prefix=prefix),
+        image_path=jammy_image_path,
+        keep_revisions=1,
+    )
+    logger.info("Uploading Noble image.")
+    store.upload_image(
+        arch=arch,
+        cloud_name=cloud_name,
+        image_name=_get_base_image_name(arch=arch, base=BaseImage.NOBLE, prefix=prefix),
+        image_path=noble_image_path,
+        keep_revisions=1,
+    )
+
+    with openstack.connect(cloud=cloud_name) as conn:
+        _create_keypair(conn=conn, prefix=prefix)
+        logger.info("Creating security group %s.", SHARED_SECURITY_GROUP_NAME)
+        _create_security_group(conn=conn)
+
+
+def _get_base_image_name(arch: Arch, base: BaseImage, prefix: str) -> str:
+    """Get formatted image name.
+
+    Args:
+        arch: The architecture of the image to use as build base.
+        base: The ubuntu base image.
+        prefix: The prefix to use for the image name.
+
+    Returns:
+        The ubuntu base image name uploaded to OpenStack.
+    """
+    return f"{prefix}-image-builder-base-{base.value}-{arch.value}"
```

Comment:
added the prefix to avoid race conditions when multiple builders use the same openstack tenant